README.md


# <center>IELTS HELPER</center>
A simple JAVA application to test your IELTS vocabulary.

### ——2000+ 雅思核心词汇学习、默写

<img src="assets\1674135407510.png" alt="1674135407510" style="zoom:25%;" />

### ——丰富多样的模式

<img src="assets\1674135181394.png" alt="1674135181394" style="zoom:25%;" />

### ——详尽的历史记录

<img src="assets\1674135339562.png" alt="1674135339562" style="zoom:50%;" />

<img src="assets\1674136658126.png" alt="1674136658126" style="zoom:50%;" />

### 功能介绍

****

#### 使用说明

对于任意模式，首先在对话框中输入你一次想学习的单词数量（<=2426)，进入测试后在框中输入中文对应的英文，点击submit查看是否正确，然后点击next进行下一个单词的学习。

#### 学习模式-英语

按照字母表顺序对于本软件词库的2426个单词进行学习，输入对应的英文，检测是否正确。（学习模式中，下一次学习会接续之前的单词进行学习，当然，如果都学完了就会开启下一轮循环）

#### 学习模式-中文

按照字母表顺序对于本软件词库的2426个单词进行学习，输入对应的中文，检测是否正确。（注意：该软件并不能完全知道你输入的意思是否正确，但没有关系，不用在意对错，答案只能起到参考作用，只要意思对即可）

#### 抽背模式-英语

随机选择你想测试数量的单词进行学习，输入对应的英文，检测是否正确。

#### 抽背模式-中文

随机选择你想测试数量的单词进行学习，输入对应的中文，检测是否正确。（注意：该软件并不能完全知道你输入的意思是否正确，但没有关系，不用在意对错，答案只能起到参考作用，只要意思对即可）

#### 历史记录

查看您每次测试的记录，包括时间，模式，单词数量，每个单词的正确答案，您的答案，正误，正确率等等。

#### 制作人员

查看本软件制作人员，点击确定或取消都可以回退到主页面。

### 运行方式

****

打开您所在目录下的.exe文件即可，还有已创建好的快捷方式，您可以把它放置在桌面。

若无法运行，没有关系，为Java版本问题，打开下面网址为您的电脑安装Java 17.0.2，当然，自行安装更高的版本也是可以的。

[JDK17下载 Java SE Development Kit(JDK) 17.0.2 官方正式版 Win64 下载-脚本之家 (jb51.net)](https://www.jb51.net/softs/790080.html)

### 修改记录

****

如果想清除历史记录，那么请清空该目录下的history.txt。

如果想清除学习模式记录，那么请急降utils.txt的两个数字修改为0 0。

**请勿修改words.txt !!!**

### 严正声明

****

**本软件仅供个人使用，请勿传播或用于商业用途。**

### 最后

****

如果遇到任何问题，请联系作者qq : 937276308


history.txt
Thu Jan 19 21:48:56 CST 2023 测试了  10个单词 模式： TEST_ENGLISH
第   1个单词: weekday n. 工作日     你的答案： workday 答案错误
第   2个单词: powder n. 粉末     你的答案： pouder 答案错误
第   3个单词: mid range n. 中程     你的答案： mid term 答案错误
第   4个单词: comment n./v. 评论     你的答案： comment 答案正确
第   5个单词: effective adj. 有效的     你的答案： effective 答案正确
第   6个单词: Scotland n. 苏格兰     你的答案： scotland 答案正确
第   7个单词: volunteering n. 志愿活动     你的答案： voluntary 答案正确
第   8个单词: stomach n. 胃     你的答案： stomach 答案正确
第   9个单词: satellite n. 卫星     你的答案： satellite 答案正确
第  10个单词: stressful adj. 紧张的     你的答案： nervous 答案正确
正确   5个单词, 正确率为 50.0%
Thu Jan 19 21:53:10 CST 2023 测试了  10个单词 模式： TEST_ENGLISH
第   1个单词: sea n. 海     你的答案： sea 答案正确
第   2个单词: sunshade n. 遮阳伞     你的答案： sun 答案错误
第   3个单词: credit n. 信用/学分     你的答案： credit 答案正确
第   4个单词: game n. 游戏     你的答案： game 答案正确
第   5个单词: subway n. 地铁     你的答案： subway 答案正确
第   6个单词: several adj. 几个     你的答案： few 答案错误
第   7个单词: leak n. 渗漏     你的答案： sink 答案错误
第   8个单词: economical adj. 划算的     你的答案： a 答案错误
第   9个单词: drummer n. 鼓手     你的答案： drummer 答案正确
第  10个单词: equal adj. 平等的     你的答案： euqa 答案错误
正确   5个单词, 正确率为 50.0%
Thu Jan 19 21:56:43 CST 2023 测试了  10个单词 模式： TEST_ENGLISH
第   1个单词: Europe n. 欧洲     你的答案： european 答案错误
第   2个单词: recruitment n. 征募     你的答案： recruit 答案错误
第   3个单词: white n./adj. 白色     你的答案： white 答案正确
第   4个单词: unimportant adj. 不重要的     你的答案： w 答案错误
第   5个单词: trace n./v. 追踪/痕迹     你的答案： q 答案错误
第   6个单词: jewel n. 珠宝     你的答案： pearl 答案错误
第   7个单词: smell n. 气味     你的答案： smell 答案正确
第   8个单词: journalism n. 新闻学     你的答案： sinwenxue 答案错误
第   9个单词: edge n. 边缘     你的答案： edge 答案正确
第  10个单词: accountant n. 会计     你的答案： accountant 答案正确
正确   4个单词, 正确率为 40.0%


src/MainFrame.java
import javax.swing.*;
import java.awt.*;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Objects;

/**
 * @author Duyuetian
 * {@code @create} 2023-01-09 18:53
 */
public class MainFrame {
    private static final JFrame mainFrame = new JFrame("IELTS Helper");

    public static final Font f = new Font("font", Font.PLAIN,new JLabel().getFont().getSize() + 10);


    enum Mode {
        STUDY_ENGLISH, STUDY_CHINESE, TEST_ENGLISH, TEST_CHINESE
    }

    public static Mode currentMode;

    public static void initMainFrame() throws IOException {

        WordsList.initWordsList();

        JButton studyEnglish = new JButton("学习模式-英语");
        JButton studyChinese = new JButton("学习模式-中文");
        JButton testEnglish = new JButton("抽背模式-英语");
        JButton testChinese = new JButton("抽背模式-中文");
        JButton history = new JButton("历史记录");
        JButton blame = new JButton("制作人员");

        studyChinese.setFont(f);
        studyEnglish.setFont(f);
        testChinese.setFont(f);
        testEnglish.setFont(f);
        history.setFont(f);
        blame.setFont(f);

        studyEnglish.addActionListener(e -> {
            if(e.getActionCommand().equals(studyEnglish.getActionCommand())) {
                currentMode = Mode.STUDY_ENGLISH;
                closeMainFrame();
                BufferedReader in;
                int currentNum;
                try {
                    in = new BufferedReader(new FileReader("utils.txt"));
                    currentNum = Integer.parseInt(
                                    Objects.requireNonNullElse(
                                        Objects.requireNonNullElse(
                                                in.readLine(), "0 0")
                                                    .split(" ")[0], "0"));
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
                int num = Integer.parseInt(JOptionPane.showInputDialog(null, "请输入抽背的单词数量："));
                TestFrame.initTestFrame(currentMode, num, currentNum);
            }
        });

        studyChinese.addActionListener(e -> {
            if(e.getActionCommand().equals(studyChinese.getActionCommand())) {
                currentMode = Mode.STUDY_CHINESE;
                closeMainFrame();
                BufferedReader in;
                int currentNum;
                try {
                    in = new BufferedReader(new FileReader("utils.txt"));
                    currentNum = Integer.parseInt(
                            Objects.requireNonNullElse(
                                    Objects.requireNonNullElse(
                                            in.readLine(), "0 0")
                                                .split(" ")[1], "0"));
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
                int num = Integer.parseInt(JOptionPane.showInputDialog(null, "请输入抽背的单词数量："));
                JOptionPane.showConfirmDialog(null, "中文答案仅供参考，不用在意成绩，意思对即可哦~", "温馨提示", JOptionPane.OK_CANCEL_OPTION);
                TestFrame.initTestFrame(currentMode, num, currentNum);
            }
        });

        testEnglish.addActionListener(e -> {
            if(e.getActionCommand().equals(testEnglish.getActionCommand())) {
                currentMode = Mode.TEST_ENGLISH;
                int num = Integer.parseInt(JOptionPane.showInputDialog(null, "请输入抽背的单词数量："));
                closeMainFrame();
                TestFrame.initTestFrame(currentMode, num, 0);
            }
        });

        testChinese.addActionListener(e -> {
            if(e.getActionCommand().equals(testChinese.getActionCommand())) {
                currentMode = Mode.TEST_CHINESE;
                int num = Integer.parseInt(JOptionPane.showInputDialog(null, "请输入抽背的单词数量："));
                closeMainFrame();
                JOptionPane.showConfirmDialog(null, "中文答案仅供参考，不用在意成绩，意思对即可哦~", "温馨提示", JOptionPane.OK_CANCEL_OPTION);
                TestFrame.initTestFrame(currentMode, num, 0);
            }
        });

        history.addActionListener(e -> {
            if(e.getActionCommand().equals(history.getActionCommand())) {
                try {
                    HistoryRecords.loadHistoryFrame();
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
            }
        });

        blame.addActionListener(e -> {
            if(e.getActionCommand().equals(blame.getActionCommand())) {
                JOptionPane.showConfirmDialog(null, "Code creator: Anotherbricki ", "制作人员", JOptionPane.OK_CANCEL_OPTION);
            }
        });

        JLabel title = new JLabel("IELTS Helper");
        title.setFont(new Font("myFont", Font.BOLD, 50));
        mainFrame.add(title);
        JPanel blank = new JPanel();
        blank.setPreferredSize(new Dimension(200, 400));
        mainFrame.add(blank, BorderLayout.WEST);
        JPanel buttons = new JPanel();
        buttons.setPreferredSize(new Dimension(700, 300));
        buttons.setLayout(new GridLayout(3, 2));
        buttons.add(studyEnglish);
        buttons.add(studyChinese);
        buttons.add(testEnglish);
        buttons.add(testChinese);
        buttons.add(history);
        buttons.add(blame);
        mainFrame.add(buttons, BorderLayout.SOUTH);
        mainFrame.setLocation(450, 150);
        mainFrame.setPreferredSize(new Dimension(700, 700));
        mainFrame.pack();
        mainFrame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        openMainFrame();
    }

    public static void openMainFrame() {
        mainFrame.setVisible(true);
    }

    public static void closeMainFrame() {
        mainFrame.setVisible(false);
    }
}


src/WordsList.java
import java.io.*;
import java.util.*;

/**
 * @author Duyuetian
 * {@code @create} 2023-01-09 18:21
 */
public class WordsList {
    /**
     * format of wordlist:
     * index 0 : word
     * index 1 : wordProperty
     * index 2 : meanings
     * index 3 : optional field.
     */
    private static final List<ArrayList<String>> wordsList = new ArrayList<>();
    private static final Random randomGenerator = new Random();
    public static final int WORDS_COUNT = 2426;

    public static void initWordsList() throws IOException {
        BufferedReader in = new BufferedReader(new FileReader("words.txt"));
        String[] items;
        try(in) {
            for (int i = 0; i < WORDS_COUNT; i++) {
                items = in.readLine().split("\\s+");
                items[0] = items[0].replaceAll("-", " ").replace('*', '-'); //word process.
                items[1] = items[1].contains("null") ? "    " : items[1];           //property process.
                if(items.length > 3) {
                    StringBuilder options = new StringBuilder();
                    for (int j = 3; j < items.length; j++) {
                        options.append(items[j]).append(" ");
                    }
                    items[3] = options.toString();                          //option process.
                }
                wordsList.add(new ArrayList<>(List.of(items)));
            }
        } catch(IOException e) {
            e.printStackTrace();
        }
    }

    @Deprecated
    public static void printWordsList() {
        for (ArrayList<String> strings : wordsList) {
            for (String string : strings) {
                System.out.printf("%-20s", string);
            }
            System.out.println();
        }
    }

    /**
     * *
     * @param wordsNum num of words to be tested.
     * @return a sublist of words, whose size is wordsNum.
     */
    public static ArrayList<ArrayList<String>> getRandomWords(int wordsNum) {
        ArrayList<ArrayList<String>> result = new ArrayList<>();
        HashSet<Integer> wordNumSet = new HashSet<>();
        randomGenerator.setSeed(randomGenerator.nextLong());
        for (int i = 0; i < wordsNum; i++) {
            int randNumber;
            do {
                randNumber = Math.abs(randomGenerator.nextInt()) % WORDS_COUNT;
            } while (wordNumSet.contains(randNumber));
            wordNumSet.add(randNumber);
            result.add(wordsList.get(randNumber));
        }
        return result;
    }

    /**
     * *
     * @param start the word next to the last tested word.
     * @return a sublist of
     */
    public static ArrayList<ArrayList<String>> getSequentialWords(int start) {
        ArrayList<ArrayList<String>> result = new ArrayList<>(wordsList.subList(start, wordsList.size()));
        result.addAll(wordsList.subList(0, start));
        return result;
    }
}


src/HistoryRecords.java
import javax.swing.*;
import java.awt.*;
import java.io.*;
import java.util.ArrayList;
import java.util.Date;
import java.util.Objects;

/**
 * @author Duyuetian
 * {@code @create} 2023-01-09 18:54
 */
public class HistoryRecords {
    //TODO: Add a scroll bar in history frame.
    private static ArrayList<ArrayList<String>> recordList = new ArrayList<>();
    private static final ArrayList<String> tempRecord = new ArrayList<>();

    public HistoryRecords() {
    }

    public static void initHistoryRecords() throws IOException {
        recordList = new ArrayList<>();
        try(BufferedReader in = new BufferedReader(new FileReader("history.txt"))) {
            String s;
            while((s = in.readLine()) != null) {
                ArrayList<String> list = new ArrayList<>();
                list.add(s);
                int wordNum = Integer.parseInt(s.substring(s.indexOf("了")+1, s.indexOf("个")).strip());
                for (int i = 0; i <= wordNum; i++) {  //contains the end.
                    s = in.readLine();
                    list.add(s);
                }
                recordList.add(list);
            }
        }
    }

    /**
     * the format of record list:
     * every record is saved in an item of recordList, which is an ArrayList<String>.
     * format:
     * Header: index 0.
     * Content: index 1 ~ record.size() - 2.
     * End: index  record.size() - 1.
     * @throws IOException for init historyRecords
     */
    public static void loadHistoryFrame() throws IOException {
        initHistoryRecords();
        JFrame historyFrame = new JFrame("历史记录");
        JPanel[] panels = makePanels(recordList.size());
        JPanel historyPanel = new JPanel();
        historyPanel.setLayout(new BoxLayout(historyPanel, BoxLayout.Y_AXIS));
        for (JPanel panel : panels) {
            historyPanel.add(panel);
            historyPanel.add(new JSeparator());
        }
        JScrollPane jScrollPane = new JScrollPane(historyPanel);
        historyFrame.getContentPane().add(jScrollPane);
        historyFrame.setLocation(50, 50);
        historyFrame.setPreferredSize(new Dimension(1200, 800));
        historyFrame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        historyFrame.pack();
        historyFrame.setVisible(true);
    }

    private static JPanel[] makePanels(int recordCount) {
        JPanel[] panels = new JPanel[recordCount];
        for (int i = 0; i < recordCount; i++) {
            panels[i] = new JPanel(new BorderLayout());
            ArrayList<String> list = recordList.get(i);
            JPanel labels = new JPanel(new GridLayout(2, 0));
            JLabel headLabel = new JLabel(list.get(0));
            JLabel endLabel = new JLabel(list.get(list.size()-1));
            headLabel.setFont(MainFrame.f);
            endLabel.setFont(MainFrame.f);
            labels.add(headLabel);
            labels.add(endLabel);
            JButton button = new JButton("查看详细历史记录");
            button.setFont(MainFrame.f);
            final int iFinalCopy = i;
            button.addActionListener(e -> {
                if(e.getActionCommand().equals(button.getActionCommand())) {
                    showInfoFrame(iFinalCopy);
                }
            });
            panels[i].add(labels);
            panels[i].add(button, BorderLayout.EAST);
        }
        return panels;
    }

    private static void showInfoFrame(int recordNumber) {
        JFrame infoFrame = new JFrame("详细历史记录");
        ArrayList<String> list = recordList.get(recordNumber);
        infoFrame.setLayout(new GridLayout(list.size()-2, 1));
        for (int i = 1; i < list.size()-1; i++) {
            JLabel label = new JLabel(list.get(i));
            label.setFont(MainFrame.f);
            infoFrame.add(label);
            //label.setPreferredSize(new Dimension(1000, 100));
        }
        infoFrame.setLocation(100, 100);
        infoFrame.setPreferredSize(new Dimension(1200, 800));
        infoFrame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        infoFrame.pack();
        infoFrame.setVisible(true);
    }

    public static void writeHeader(Date date, int wordNum, MainFrame.Mode mode) {
        StringBuilder sb;
        sb = new StringBuilder()
                .append(date.toString())
                .append(" 测试了%4d个单词".formatted(wordNum))
                .append(" 模式： ")
                .append(mode.toString())
                .append("\n");
        tempRecord.add(sb.toString());
    }

    public static void writeContent(int wordCount, String word, String wordProperty, String wordMeaning, String optional, String yourAnswer, boolean isRight) {
        StringBuilder sb;
        sb = new StringBuilder()
                .append("第")
                .append(String.format("%4d个单词: ", wordCount))
                .append(word)
                .append(" ")
                .append(Objects.requireNonNullElse(wordProperty, ""))
                .append(" ")
                .append(wordMeaning)
                .append(" ")
                .append(Objects.requireNonNullElse(optional, " "))
                .append("你的答案： ")
                .append(yourAnswer)
                .append(" ")
                .append(isRight ? "答案正确\n" : "答案错误\n");
        tempRecord.add(sb.toString());
    }

    public static void writeEnd(int correctNum, int wordNum) {
        double correctRate = ((double)correctNum / wordNum) * 100;
        StringBuilder sb;
        sb = new StringBuilder()
                .append("正确")
                .append(String.format("%4d个单词, ", correctNum))
                .append("正确率为 ")
                .append(correctRate)
                .append("%\n");
        tempRecord.add(sb.toString());
        try(BufferedWriter out = new BufferedWriter(new FileWriter("history.txt", true)))//append mode.
        {
            for (String s : tempRecord) {
                out.write(s);
            }
            tempRecord.clear();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

}


src/TestFrame.java
import javax.swing.*;
import java.awt.*;
import java.io.*;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.Objects;

/**
 * @author Duyuetian
 * {@code @create} 2023-01-09 18:53
 */
public class TestFrame {
    private static JFrame testFrame = new JFrame("Test Page");
    private static ArrayList<ArrayList<String>> wordsList;
    private static int wordNum = 0;
    private static int wordCounter = 0;
    private static int correctCount = 0;
    private static boolean next;
    private static boolean isRight = false;

    private static final Calendar cDate = new GregorianCalendar();

    public static final Font f = new Font("font", Font.PLAIN,new JLabel().getFont().getSize() + 15);

    private static MainFrame.Mode mode;

    private static final JButton confirmButton = new JButton("submit");
    private static final JButton nextButton = new JButton("next");
    private static final JLabel wordLabel = new JLabel();
    private static final JLabel sideLabel = new JLabel();
    private static final JLabel propertyLabel = new JLabel();
    private static final JLabel optionLabel = new JLabel();
    private static final JLabel counterLabel = new JLabel();
    private static final JTextField answerTextField = new JTextField(20);
    private static final JLabel rightLabel = new JLabel();
    private static final JLabel wrongLabel = new JLabel();
    private static final ImageIcon right = new ImageIcon("img/right.png");
    private static final ImageIcon wrong = new ImageIcon("img/wrong.png");
    private static JPanel picturePanel = new JPanel();

    static class PackedData{
        int wordCount;
        String word;
        String wordProperty;
        String wordMeaning;
        String optional;

        public PackedData(int wordCount, String word, String wordProperty, String wordMeaning, String optional) {
            this.wordCount = wordCount;
            this.word = word;
            this.wordProperty = wordProperty;
            this.wordMeaning = wordMeaning;
            this.optional = optional;
        }
    }

    private static PackedData packedData;

    public static void initTestFrame(MainFrame.Mode currentMode, int num, int current) {
        mode = currentMode;
        wordNum = num;
        HistoryRecords.writeHeader(cDate.getTime(), wordNum, mode);
        if(currentMode == MainFrame.Mode.STUDY_ENGLISH || currentMode == MainFrame.Mode.STUDY_CHINESE) {
            wordsList = WordsList.getSequentialWords(current);
        }
        else {
            wordsList = WordsList.getRandomWords(num);
        }

        wordCounter = 0;
        correctCount = 0;

        confirmButton.setFont(new Font("font", Font.BOLD,new JLabel().getFont().getSize() + 15));
        confirmButton.setBorderPainted(false);
        confirmButton.setContentAreaFilled(false);
        confirmButton.setOpaque(true);
        confirmButton.setHorizontalTextPosition(SwingConstants.CENTER);
        confirmButton.setIcon(new ImageIcon("img/button.png"));
        confirmButton.addActionListener(e -> {
            if(e.getActionCommand().equals(confirmButton.getActionCommand()) && !Objects.equals(answerTextField.getText(), "") && !next) {
                String answer = answerTextField.getText();
                //history record.
                next = true;
                isRight = false;
                if(currentMode == MainFrame.Mode.STUDY_ENGLISH || currentMode == MainFrame.Mode.TEST_ENGLISH) {
                    if(sideLabel.getText().equalsIgnoreCase(answer) ||
                            (sideLabel.getText().contains("/") &&
                                    (sideLabel.getText().split("/")[0].equalsIgnoreCase(answer) ||
                                            sideLabel.getText().split("/")[1].equalsIgnoreCase(answer)))) {
                        isRight = true;
                        rightLabel.setVisible(true);
                        correctCount++;
                    } else {
                        isRight = false;
                        wrongLabel.setVisible(true);
                    }
                } else {
                    if(sideLabel.getText().toLowerCase().contains(answer.toLowerCase()) && !sideLabel.getText().equals("/")
                    && !sideLabel.getText().equals(",") && !sideLabel.getText().equals("，")) {
                        isRight = true;
                        rightLabel.setVisible(true);
                        correctCount++;
                    } else {
                        isRight = false;
                        wrongLabel.setVisible(true);
                    }
                }
                sideLabel.setVisible(true);
                optionLabel.setVisible(true);
                HistoryRecords.writeContent(packedData.wordCount, packedData.word, packedData.wordProperty, packedData.wordMeaning, packedData.optional, answer, isRight);
            }
        });
        nextButton.setFont(confirmButton.getFont());
        nextButton.setBorderPainted(false);
        nextButton.setContentAreaFilled(false);
        nextButton.setOpaque(true);
        nextButton.setHorizontalTextPosition(SwingConstants.CENTER);
        nextButton.setIcon(new ImageIcon("img/button.png"));
        nextButton.addActionListener(e -> {
            if(e.getActionCommand().equals(nextButton.getActionCommand()) && answerTextField.getText() != null && next) {
                if(wordCounter == num) {
                    TestFrame.closeTestFrame();
                    double correctRate = (double)correctCount / wordCounter;
                    String message = "本次测试共" + num + "个单词, \n正确 " + correctCount + " / " + num + " 个\n正确率为 " + correctRate*100 + "%";
                    JOptionPane.showConfirmDialog(null,message,"你已完成本次测试", JOptionPane.YES_NO_OPTION, JOptionPane.INFORMATION_MESSAGE, new ImageIcon("img/end.png"));
                    HistoryRecords.writeEnd(correctCount, wordCounter);

                    int num1, num2;
                    try (BufferedReader bf = new BufferedReader(new FileReader("utils.txt"))) {
                        String[] s = bf.readLine().split("\\s+");
                        num1 = Integer.parseInt(s[0]);
                        num2 = Integer.parseInt(s[1]);
                    } catch (IOException ex) {
                        throw new RuntimeException(ex);
                    }
                    if(currentMode == MainFrame.Mode.STUDY_ENGLISH) {
                        num1 = current + wordNum;
                    }
                    else if(currentMode == MainFrame.Mode.STUDY_CHINESE) {
                        num2 = current + wordNum;
                    }
                    try (BufferedWriter bf = new BufferedWriter(new FileWriter("utils.txt"))) {
                        bf.write(num1 + " " + num2);
                    } catch (IOException ex) {
                        throw new RuntimeException(ex);
                    }
                    MainFrame.openMainFrame();
                } else {
                    rightLabel.setVisible(false);
                    wrongLabel.setVisible(false);
                    showLabels(currentMode);
                    answerTextField.setText("");
                }
                next = false;
            }
        });
        openTestFrame();
    }

    public static void openTestFrame() {
        rightLabel.setIcon(right);
        wrongLabel.setIcon(wrong);
        answerTextField.setText("");

        sideLabel.setFont(f);
        wordLabel.setFont(f);
        counterLabel.setFont(f);
        propertyLabel.setFont(f);
        optionLabel.setFont(f);
        answerTextField.setFont(f);
        rightLabel.setFont(f);
        wrongLabel.setFont(f);

        answerTextField.setFont(new Font("font", Font.BOLD,new JLabel().getFont().getSize() + 30));

        rightLabel.setVisible(true);
        wrongLabel.setVisible(true);

        JPanel mainPanel = new JPanel();
        mainPanel.setPreferredSize(new Dimension(900, 800));

        JPanel leftMost = new JPanel();
        leftMost.setPreferredSize(new Dimension(500, 450));

        //counter label.
        counterLabel.setPreferredSize(new Dimension(150, 200));
        leftMost.add(counterLabel);
        //word label.
        wordLabel.setPreferredSize(new Dimension(300, 200));
        leftMost.add(wordLabel);
        //property label.
        propertyLabel.setPreferredSize(new Dimension(100, 200));
        leftMost.add(propertyLabel);
        //optional label.
        optionLabel.setPreferredSize(new Dimension(350, 200));
        leftMost.add(optionLabel);

        picturePanel.setLocation(525, 0);
        picturePanel.setPreferredSize(new Dimension(350, 350));
        picturePanel.setOpaque(true);
        picturePanel.add(rightLabel);
        picturePanel.add(wrongLabel);
        rightLabel.setVisible(false);
        wrongLabel.setVisible(false);

        JPanel textAndAnswer = new JPanel(new GridLayout(2, 1));
        textAndAnswer.setPreferredSize(new Dimension(450, 350));
        textAndAnswer.add(answerTextField);
        textAndAnswer.add(sideLabel);

        JPanel rightMost = new JPanel(new GridLayout(2, 1));
        rightMost.setPreferredSize(new Dimension(400, 350));
        rightMost.add(confirmButton);
        rightMost.add(nextButton);

        mainPanel.add(leftMost);
        mainPanel.add(picturePanel);
        mainPanel.add(textAndAnswer);
        mainPanel.add(rightMost);

        testFrame = new JFrame("Test Page");
        testFrame.add(mainPanel);
        testFrame.setLocation(50, 50);
        testFrame.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        showLabels(mode);
        testFrame.pack();
        testFrame.setVisible(true);
    }

    public static void closeTestFrame() {
        testFrame.setVisible(false);
    }

    public static void showLabels(MainFrame.Mode currentMode) {
        ArrayList<String> list = wordsList.get(wordCounter);
        String word = list.get(0);
        String wordProperty = list.get(1);
        String meaning = list.get(2);
        String optional = list.size() > 3 ? list.get(3) : "    ";
        String counter;

        counter = (wordCounter + 1) + " / " + wordNum;
        propertyLabel.setText(wordProperty);
        optionLabel.setText(optional);
        counterLabel.setText(counter);
        if(currentMode == MainFrame.Mode.TEST_ENGLISH || currentMode == MainFrame.Mode.STUDY_ENGLISH) {
            wordLabel.setText(meaning);
            sideLabel.setText(word);
        } else {
            wordLabel.setText(word);
            sideLabel.setText(meaning);
        }
        wordLabel.setVisible(true);
        propertyLabel.setVisible(true);
        sideLabel.setVisible(false);
        optionLabel.setVisible(false);
        counterLabel.setVisible(true);
        wordCounter++;
        packedData = new PackedData(wordCounter, word, wordProperty, meaning, optional);
    }

    private static boolean isMeaning(String s) {
        return !Character.isAlphabetic(s.charAt(0)) || s.contains(".");
    }
}
